### PR TITLE
translate the User input for parity to the application

### PR DESCRIPTION
**Pull request by GhasBarbouchi**
.update main.py
.update Module.yaml
.update requirement.txt
.update application.py

### CODE REVIEW

File: Module.yaml [hunk 1]
```diff
@@ -49,15 +49,14 @@ environments:
...
-      - 'M'
-      - 'S'
+
+    default: 'None'
+    type: string
     name: Parity
+    options:
+    - 'None'
+    - 'Even'
+    - 'Odd'
```
Comment (matrixtwister) [inline Module.yaml:59]: what about the other two values?

File: image/src/app/config/application.py [hunk 2]
```diff
@@ -1,12 +1,11 @@
...
 """
-import serial
 from app.utils.env import env
 APPLICATION = {
     "PORT":      env("PORT","/dev/ttyUSB0"),
     "BAUD_RATE": env("BAUD_RATE","115200"),
-    "DATA_BITS": env("DATA_BITS",serial.EIGHTBITS),
-    "PARITY":    env("PARITY",serial.PARITY_NONE),
-    "STOP_BITS": env("STOP_BITS",serial.STOPBITS_ONE)
+    "DATA_BITS": env("DATA_BITS",8),
```
Comment (matrixtwister) [inline image/src/app/config/application.py:8]: why not use a named constant here? that never hurts

File: image/src/app/config/application.py [hunk 3]
```diff
@@ -1,12 +1,11 @@
...
+    "PARITY":    env("PARITY",'None'),
+    "STOP_BITS": env("STOP_BITS",1)
```
Comment (matrixtwister) [inline image/src/app/config/application.py:10]: why not use a named constant here? that never hurts

File: image/src/app/module/main.py [hunk 4]
```diff
@@ -17,8 +25,8 @@ def module_main():
     """
     try:
         #  open the serial port and get the serial port object
-        ser = serial.Serial(APPLICATION['PORT'], APPLICATION['BAUD_RATE'], timeout=1,bytesize=APPLICATION['DATA_BITS'],parity=APPLICATION['PARITY'], stopbits=APPLICATION['STOP_BITS'])
-        while True: 
+        ser = serial.Serial(APPLICATION['PORT'], APPLICATION['BAUD_RATE'], timeout=1,bytesize=APPLICATION['DATA_BITS'],parity=switcher.get(sParity, lambda: "Invalid parity"), stopbits=APPLICATION['STOP_BITS'])
+        while True:
           #  read json payload
           ser.reset_input_buffer()
           data = ser.readline().decode("ISO-8859-1")
```
Comment (matrixtwister) [inline image/src/app/module/main.py:32]: the indentation is inconsistent here

**Review by matrixtwister — CHANGES_REQUESTED** [review]

File: Module.yaml (hunk 1 above)
Comment (GhasBarbouchi) [reply Module.yaml:58]: we will not use the 9 bits so , it does not matter to use “Mark/Space” Christos informed is not required.

File: image/src/app/config/application.py (hunk 2 above)
Comment (GhasBarbouchi) [reply image/src/app/config/application.py:8]: because using the switch_demo which translates the variable environment from application.py

Comment (matrixtwister) [reply image/src/app/config/application.py:8]: the comment is related to line `8`, so `DATA_BITS`

File: image/src/app/config/application.py [hunk 5]
```diff
@@ -1,12 +1,11 @@
...
 """
-import serial
 from app.utils.env import env
 APPLICATION = {
     "PORT":      env("PORT","/dev/ttyUSB0"),
     "BAUD_RATE": env("BAUD_RATE","115200"),
-    "DATA_BITS": env("DATA_BITS",serial.EIGHTBITS),
-    "PARITY":    env("PARITY",serial.PARITY_NONE),
-    "STOP_BITS": env("STOP_BITS",serial.STOPBITS_ONE)
+    "DATA_BITS": env("DATA_BITS","EIGHTBITS"),
```
Comment (matrixtwister) [inline image/src/app/config/application.py:8]: please change it back to the named constant

File: image/src/app/module/main.py [hunk 6]
```diff
@@ -6,6 +6,15 @@
 import json
 from app.weeve.egress import send_data
 from app.config import APPLICATION
+
+sParity=APPLICATION['PARITY']
+def switch_demo(sParity):
```
Comment (matrixtwister) [inline image/src/app/module/main.py:11]: please choose a more meaningful name for this function

File: image/src/app/module/main.py [hunk 7]
```diff
@@ -6,6 +6,14 @@
 import json
 from app.weeve.egress import send_data
 from app.config import APPLICATION
+
+def parity_translation(sParity=APPLICATION['PARITY']):
+    switcher = {
+       "None" : "N",
+       "Even" : "E",
+        "Odd" : "O"
```
Comment (matrixtwister) [inline image/src/app/module/main.py:14]: you can also used name constants for `"N"`, `"E"` and `"O"`

**Review by matrixtwister — CHANGES_REQUESTED** [review]

File: config.env [hunk 8]
```diff
@@ -7,8 +7,7 @@ EGRESS_URL=http://0.0.0.0:80
 PORT=/dev/ttyUSB0
 BAUD_RATE=115200
 DATA_BITS=8
-PARITY='N'
+PARITY="N"
```
Comment (matrixtwister) [inline config.env:10]: this should be `'None'`

File: image/src/app/module/main.py [hunk 9]
```diff
@@ -15,18 +16,26 @@ def module_main():
...
         [string, string]: [data, error]
     """
+    parity_dict = {
+       "None" : serial.PARITY_NONE,
+       "Even" : serial.PARITY_EVEN,
+       "Odd"  : serial.PARITY_ODD
+    }
     try:
+        if APPLICATION['PARITY'] not in parity_dict:
+            raise Exception("..Parity not validated..")
```
Comment (matrixtwister) [inline image/src/app/module/main.py:26]:
```suggestion
            raise Exception("Invalid parity")
```

File: image/src/app/module/main.py [hunk 10]
```diff
@@ -15,18 +16,26 @@ def module_main():
     Returns:
         [string, string]: [data, error]
     """
+    parity_dict = {
+       "None" : serial.PARITY_NONE,
+       "Even" : serial.PARITY_EVEN,
+       "Odd"  : serial.PARITY_ODD
+    }
     try:
```
Comment (matrixtwister) [inline image/src/app/module/main.py:24]: it makes no sense to raise an exception in the try-except block

**Review by matrixtwister — CHANGES_REQUESTED** [review]

**Review by GhasBarbouchi — COMMENTED** [review]
convert stop bits to float

File: image/src/app/module/main.py [hunk 11]
```diff
@@ -15,18 +16,30 @@ def module_main():
...
-          data = ser.readline().decode("ISO-8859-1")
-          dict_json = json.loads(data)
-          sent=send_data(dict_json)
-          if sent:
-            print("SUCCESS"),print("failed")
-          print(dict_json)
-    except json.JSONDecodeError as e:
-      return None, f"Unable to perform the module logic: {e}"
-      
+        str_data_bytes = bytes(str(APPLICATION['DATA_BITS']), "ascii")
```
Comment (matrixtwister) [inline image/src/app/module/main.py:28]: the intermediary step with bytes seems unnecessary

File: image/src/app/module/main.py [hunk 12]
```diff
@@ -15,18 +16,30 @@ def module_main():
...
+        data_bytes= int(str_data_bytes)
+        str_stop_bits=bytes(str(APPLICATION['STOP_BITS']), "ascii")
```
Comment (matrixtwister) [inline image/src/app/module/main.py:30]: the intermediary step with bytes seems unnecessary

**Review by matrixtwister — APPROVED** [review]
@GhasBarbouchi can be merged after the rest of discussions is resolved

File: image/src/app/module/main.py (hunk 11 above)
Comment (GhasBarbouchi) [reply image/src/app/module/main.py:28]: I didn't update the code yet, but yes it is unnecessary , just thinking and coding without removing the unessassary

File: image/src/app/module/main.py (hunk 12 above)
Comment (GhasBarbouchi) [reply image/src/app/module/main.py:30]: yes it is unnessasary

**Pull request merged**